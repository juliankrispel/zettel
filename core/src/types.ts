/**
 * Container for Entities
 */
export type EntityMap = {
  [key: string]: Entity
}

/**
 * Entity ftw!
 */
export type Entity = {
  isAtomic?: boolean,
  [key: string]: any
}

/**
 * State of a document.
 * Used to create derived states such as BlockTree,
 * RawDocument or any other
 */
export type ListState = {
  value: Value,
  entityMap: EntityMap,
}

/**
 * Array of Characters
*/
export type Value = Character[]

export type Character = (TextCharacter | BlockStart | BlockEnd)

export type CharacterData = {
  styles: string[],
  entity?: string | null,
  type?: void,
}

/*
* Represents one text character.
* Contains text symbol as well as array of
* styles and entity
*/
export type TextCharacter = CharacterData & {
  char: string,
}

/**
 * Represents the beginning of a block
 */
export type BlockStart = {
  blockKey: string,
  type: 'block-start',
  entity?: string | null
}

export type TextFragment = CharacterData & {
  text: string,
  entity?: Entity
}

/**
 * Represents the end of a block, there
 * has toe be a matching 
 */
export type BlockEnd = {
  type: 'block-end'
}

/**
 * Embodies all possible changes made to doc
 */
export type Change = {
  start: number,
  end: number,
  value: Value | CharacterData
}

/**
 * History
 */
export type Changes = Change[]

/**
 * Represents raw form of a Zettel Document,
 * serializable and easier on the human eye
 */
export type RawDocument = {
  text: string,
  ranges: RawRange[],
  entityMap: EntityMap,
}

/**
 * Ranges in RawDocument
 * Used to style text and relate to entities
 */
export type RawRange = {
  offset: number,
  length: number,
  styles: string[],
  entity?: string | null,
}

export type CharacterRange = RawRange

/**
* Tree representation of content. Used for
* rendering only
*/
export type BlockTree = {
  blocks: Block[],
  entityMap: EntityMap,
}

/**
* Represents one node in a tree
* Used for rendering
*/
export type Block = {
  value: TextCharacter[],
  blockKey: string,
  blocks: Block[],
  entity?: Entity | null
}


/**
 * 
 */
export interface EditorState {
  /**
   * @method change()
   * 
   * `EditorState.change({ start: 9, end: 18, value: [] })`
   */
  change: (changes: Change[]) => EditorState;

  /**
   * @method undo()
   * 
   * `EditorState.undo()`
   */
  undo: () => EditorState;

  /**
   * @method redo()
   * 
   * `EditorState.redo()`
   */
  redo: () => EditorState;

  readonly list: ListState;
  readonly tree: BlockTree;
}