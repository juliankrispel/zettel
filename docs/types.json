[
  {
    "file": "/Users/juliankrispel/projects/zettel-new/react/src/types.ts",
    "nodes": [
      {
        "docComment": "Render",
        "kind": "TypeAliasDeclaration",
        "name": "RenderStyle",
        "type": "React.FunctionComponent<{\n  style: string,\n  children: React.ReactElement\n}>"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/react/src/Editor.tsx",
    "nodes": [
      {
        "name": null,
        "kind": "VariableStatement",
        "docComment": "Editor Component",
        "parameters": [
          {
            "isOptional": false,
            "name": "props",
            "type": {
              "mapBlock": "(block: Block) => Block",
              "renderBlock": "RenderBlock",
              "renderChildren": "RenderChildren",
              "renderStyle": "RenderStyle",
              "renderTextFragment": "RenderTextFragment",
              "onChange": "(editorState: EditorState) => void",
              "editorState": "EditorState",
              "htmlAttrs": "Object",
              "onKeyDown": "(event: React.KeyboardEvent) => EditorState | void",
              "readOnly": "boolean"
            }
          }
        ],
        "type": "React.ReactElement"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/types.ts",
    "nodes": [
      {
        "docComment": "Container for Entities",
        "kind": "TypeAliasDeclaration",
        "name": "EntityMap",
        "type": {}
      },
      {
        "docComment": "Entity ftw!",
        "kind": "TypeAliasDeclaration",
        "name": "Entity",
        "type": {}
      },
      {
        "docComment": "State of a document.\nUsed to create derived states such as BlockTree,\nRawDocument or any other",
        "kind": "TypeAliasDeclaration",
        "name": "ListState",
        "type": {}
      },
      {
        "docComment": "Array of Characters",
        "kind": "TypeAliasDeclaration",
        "name": "Value",
        "type": "Character[]"
      },
      {
        "docComment": "Represents the beginning of a block",
        "kind": "TypeAliasDeclaration",
        "name": "BlockStart",
        "type": {}
      },
      {
        "docComment": "Represents the end of a block, there\nhas toe be a matching",
        "kind": "TypeAliasDeclaration",
        "name": "BlockEnd",
        "type": {}
      },
      {
        "docComment": "Embodies all possible changes made to doc",
        "kind": "TypeAliasDeclaration",
        "name": "Change",
        "type": {}
      },
      {
        "docComment": "History",
        "kind": "TypeAliasDeclaration",
        "name": "Changes",
        "type": "Change[]"
      },
      {
        "docComment": "Represents raw form of a Zettel Document,\nserializable and easier on the human eye",
        "kind": "TypeAliasDeclaration",
        "name": "RawDocument",
        "type": {}
      },
      {
        "docComment": "Ranges in RawDocument\nUsed to style text and relate to entities",
        "kind": "TypeAliasDeclaration",
        "name": "RawRange",
        "type": {}
      },
      {
        "docComment": "Tree representation of content. Used for\nrendering only",
        "kind": "TypeAliasDeclaration",
        "name": "BlockTree",
        "type": {}
      },
      {
        "docComment": "Represents one node in a tree\nUsed for rendering",
        "kind": "TypeAliasDeclaration",
        "name": "Block",
        "type": {}
      },
      {
        "kind": "InterfaceDeclaration",
        "text": "export interface EditorState {\n  /**\n   * @method change()\n   * \n   * `EditorState.change({ start: 9, end: 18, value: [] })`\n   */\n  change: (changes: Change[]) => EditorState;\n\n  /**\n   * @method undo()\n   * \n   * `EditorState.undo()`\n   */\n  undo: () => EditorState;\n\n  /**\n   * @method redo()\n   * \n   * `EditorState.redo()`\n   */\n  redo: () => EditorState;\n\n  readonly list: ListState;\n  readonly tree: BlockTree;\n}"
      },
      {
        "kind": "PropertySignature",
        "text": "change: (changes: Change[]) => EditorState;"
      },
      {
        "kind": "PropertySignature",
        "text": "undo: () => EditorState;"
      },
      {
        "kind": "PropertySignature",
        "text": "redo: () => EditorState;"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/EditorState/change.ts",
    "nodes": [
      {
        "docComment": "Represents the input and output for updates",
        "kind": "TypeAliasDeclaration",
        "name": "Update",
        "type": {}
      },
      {
        "name": "change",
        "kind": "FunctionDeclaration",
        "docComment": "takes State and Change object, return a state\nobject that has been updated with the Change, and\na Change object to undo the change",
        "type": "Update",
        "parameters": [
          {
            "isOptional": false,
            "name": "update",
            "type": "Update"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/EditorState/EditorState.ts",
    "nodes": [
      {
        "name": "getSelectedValue",
        "kind": "MethodDeclaration",
        "docComment": "gets selected value, arguments default\nto editorState.start and editorState.end",
        "type": "Value",
        "parameters": [
          {
            "isOptional": false,
            "name": "start",
            "type": "number"
          },
          {
            "isOptional": false,
            "name": "end",
            "type": "number"
          }
        ]
      },
      {
        "name": "fromJSON",
        "kind": "MethodDeclaration",
        "docComment": "creates a new EditorState from JSON format",
        "type": "EditorState",
        "parameters": [
          {
            "isOptional": false,
            "name": "json",
            "type": "RawDocument"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/getBlocksForRange.ts",
    "nodes": [
      {
        "name": "getBlocksForRange",
        "kind": "FunctionDeclaration",
        "docComment": "Gets block indeces for range (start, end)",
        "type": "BlockAndIndex[]",
        "parameters": [
          {
            "isOptional": false,
            "name": "value",
            "type": "Value"
          },
          {
            "isOptional": false,
            "name": "start",
            "type": "number"
          },
          {
            "isOptional": false,
            "name": "end",
            "type": "number"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/textToListIndex.ts",
    "nodes": [
      {
        "name": "textToListIndex",
        "kind": "FunctionDeclaration",
        "docComment": "Maps text index to list index",
        "type": "number",
        "parameters": [
          {
            "isOptional": false,
            "name": "value",
            "type": "Value"
          },
          {
            "isOptional": false,
            "name": "textIndex",
            "type": "number"
          }
        ]
      }
    ]
  }
]