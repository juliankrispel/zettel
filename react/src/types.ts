import { TextFragment, Block } from '@zettel/core'
import { CSSProperties } from 'react';

type FragmentRenderProps = {
  fragment: TextFragment
}

export type RenderStyle = React.FunctionComponent<{
  style: string,
  children: React.ReactElement
}>

export type RenderEntity = React.FunctionComponent<{
  entity: any,
  children: React.ReactElement
}>

export type RenderBlock = React.FunctionComponent<{
  block: Block,
  style: CSSProperties,
  children: React.ReactElement
}>

export type RenderChildren = React.FunctionComponent<{
  block?: Block,
  children: React.ReactElement
}>

export type RenderProps = {
  mapBlock?: (block: Block) => Block,
  renderBlock?: RenderBlock,
  renderChildren?: RenderChildren,
  renderStyle?: RenderStyle,
  renderEntity?: RenderEntity,
}  