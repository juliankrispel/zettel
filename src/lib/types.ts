/**
 * Container for Entities
 */
export type EntityMap = {
  [key: string]: any
}

/**
 * State of a document.
 * Used to create derived states such as BlockTree,
 * RawDocument or any other
 */
export type ListState = {
  value: Value,
  entityMap: EntityMap,
}

/**
* Value object. Think of this as your `value` in `<input value={value} />`,
* except instead of a string you get an array of objects.
*/
export type Value = (Character | BlockStart | BlockEnd)[]

/*
* Represents one text character.
* Contains text symbol as well as array of
* styles and entity
*/
export type Character = {
  char: string,
  styles: string[],
  entity?: string | null,
  type?: void,
}

/**
 * Represents the beginning of a block
 */
export type BlockStart = {
  type: 'block-start',
  data?: any
}

/**
 * Represents the end of a block, there
 * has toe be a matching 
 */
export type BlockEnd = {
  type: 'block-end'
}

/**
 * Embodies all possible changes made to doc
 */
export type Change = {
  start: number,
  end: number,
  value: Value
}[]

/**
 * Represents raw form of a Zettel Document,
 * serializable and easier on the human eye
 */
export type RawDocument = {
  text: string,
  ranges: RawRange[],
  entityMap: EntityMap,
}

/**
 * Ranges in RawDocument
 * Used to style text and relate to entities
 */
export type RawRange = {
  offset: number,
  length: number,
  styles: string[],
  entity?: string | null,
}

/**
* Tree representation of content. Used for
* rendering only
*/
export type BlockTree = {
  blocks: Block[],
  entityMap: EntityMap,
}

/**
* Represents one node in a tree
* Used for rendering
*/
export type Block = {
  value: Character[],
  key: string,
  blocks: Block[],
}


/**
 * 
 */
export interface EditorState {
  /**
   * @method change()
   * 
   * `EditorState.change({ start: 9, end: 18, value: [] })`
   */
  change: (change: Change) => EditorState;

  /**
   * @method undo()
   * 
   * `EditorState.undo()`
   */
  undo: () => EditorState;

  /**
   * @method redo()
   * 
   * `EditorState.redo()`
   */
  redo: () => EditorState;

  readonly list: ListState;
  readonly tree: BlockTree;
}