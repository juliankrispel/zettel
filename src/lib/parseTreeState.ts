import { FlatState, TreeState, Block } from './types'
import id from './id'

const getNode = (state: TreeState, path: number[]): Block  => {
  return path.reduce((acc: any, val) => {
    return acc.nodes[val]
  }, state)
}


const getNodes = (state: TreeState, path: number[]): Block[]  => {
  if (path.length === 0) {
    return state.nodes
  } else {
    return path.reduce((acc, val) => {
      return acc[val].nodes || acc
    }, state.nodes)
  }
}

const parseTreeState = (flat: FlatState): TreeState => {
  const state: TreeState = {
    nodes: [],
    meta: {},
  }

  let path: number[] = []

  flat.value.forEach((char, index) => {
    // if we have a start block character,
    // create a new block and add it
    if (char.char === '[') {
      const nodes = getNodes(state, path)

      nodes.push({
        value: [],
        nodes: [],
        key: id(),
      })

      path.push(nodes.length - 1)
    } else if (char.char === ']') {
      path.pop()
    } else {
      const node = getNode(state, path) || state
      node.value.push({
        char: char.char,
        styles: []
      })
    }
  })

  return state
}

export default parseTreeState