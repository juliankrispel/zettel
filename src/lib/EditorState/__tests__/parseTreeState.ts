import parseTreeState from '../flatToTree'
import { ListState } from '../../types'

jest.mock('../id', () => {
  var i = 0
  return () => `key-${i++}`
})

describe('parseTreeState', () => {
  test('parses simple flat structure', () => {
    const flatState: ListState = {
      value: [
        { type: 'block-start' },
        { char: 'H', styles: [] },
        { char: 'e', styles: [] },
        { char: 'l', styles: [] },
        { char: 'l', styles: [] },
        { char: 'o', styles: [] },
        { type: 'block-end' },
      ],
      entityMap: {}
    }

    expect(parseTreeState(flatState)).toMatchSnapshot()
  })

  test('parses nested structure', () => {
    const flatState: ListState = {
      value: [
        { type: 'block-start' },
        { char: '1', styles: [] },
        { type: 'block-start' },
        { char: '2', styles: [] },
        { type: 'block-start' },
        { char: '3', styles: [] },
        { type: 'block-end' },
        { type: 'block-start' },
        { char: '4', styles: [] },
        { type: 'block-end' },
        { type: 'block-end' },
        { type: 'block-end' },
      ],
      entityMap: {}
    }

    expect(parseTreeState(flatState)).toMatchSnapshot()
  })

  test('parses invalid structure', () => {
    const flatState: ListState = {
      value: [
        { type: 'block-start' },
        { char: 'H', styles: [] },
        { type: 'block-start' },
        { char: 'l', styles: [] },
        { char: 'l', styles: [] },
        { char: 'o', styles: [] },
        { type: 'block-end' },
      ],
      entityMap: {}
    }

    expect(parseTreeState(flatState)).toMatchSnapshot()
  })
})