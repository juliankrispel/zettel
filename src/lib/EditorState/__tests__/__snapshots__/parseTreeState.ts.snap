// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTreeState parses invalid structure 1`] = `
Object {
  "blocks": Array [
    Object {
      "blocks": Array [
        Object {
          "blocks": Array [],
          "key": "key-6",
          "value": Array [
            Object {
              "char": "l",
              "styles": Array [],
            },
            Object {
              "char": "l",
              "styles": Array [],
            },
            Object {
              "char": "o",
              "styles": Array [],
            },
          ],
        },
      ],
      "key": "key-5",
      "value": Array [
        Object {
          "char": "H",
          "styles": Array [],
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`parseTreeState parses nested structure 1`] = `
Object {
  "blocks": Array [
    Object {
      "blocks": Array [
        Object {
          "blocks": Array [
            Object {
              "blocks": Array [],
              "key": "key-3",
              "value": Array [
                Object {
                  "char": "3",
                  "styles": Array [],
                },
              ],
            },
            Object {
              "blocks": Array [],
              "key": "key-4",
              "value": Array [
                Object {
                  "char": "4",
                  "styles": Array [],
                },
              ],
            },
          ],
          "key": "key-2",
          "value": Array [
            Object {
              "char": "2",
              "styles": Array [],
            },
          ],
        },
      ],
      "key": "key-1",
      "value": Array [
        Object {
          "char": "1",
          "styles": Array [],
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`parseTreeState parses simple flat structure 1`] = `
Object {
  "blocks": Array [
    Object {
      "blocks": Array [],
      "key": "key-0",
      "value": Array [
        Object {
          "char": "H",
          "styles": Array [],
        },
        Object {
          "char": "e",
          "styles": Array [],
        },
        Object {
          "char": "l",
          "styles": Array [],
        },
        Object {
          "char": "l",
          "styles": Array [],
        },
        Object {
          "char": "o",
          "styles": Array [],
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;
